File: src/gmsh_backend.cpp
#include "stator/gmsh_backend.hpp"
#include <iostream>
#include <stdexcept>

namespace stator {

// ─── StubGmshBackend ─────────────────────────────────────────────────────────

void StubGmshBackend::reset() {
    initialized_      = false;
    sync_count_       = 0;
    finalized_        = false;
    mesh_generated_   = false;
    background_field_ = -1;
    last_write_path_.clear();
    point_counter_        = 0;
    line_counter_         = 0;
    surface_counter_      = 0;
    curve_loop_counter_   = 0;
    field_counter_        = 0;
    phys_group_tag_counter_ = 1000;
    phys_groups_.clear();
    surfaces_2d_.clear();
}

void StubGmshBackend::initialize(const std::string& /*model_name*/) {
    initialized_ = true;
}

void StubGmshBackend::synchronize() { ++sync_count_; }

void StubGmshBackend::finalize() { finalized_ = true; }

void StubGmshBackend::set_option(const std::string& /*name*/, double /*value*/) {}

int StubGmshBackend::add_point(double, double, double, double) {
    return ++point_counter_;
}

int StubGmshBackend::add_line(int, int) { return ++line_counter_; }

int StubGmshBackend::add_circle(double, double, double, double) {
    return ++line_counter_;
}

int StubGmshBackend::add_arc(int, int, int) { return ++line_counter_; }

int StubGmshBackend::add_curve_loop(const std::vector<int>&) {
    return ++curve_loop_counter_;
}

int StubGmshBackend::add_plane_surface(const std::vector<int>&) {
    int tag = ++surface_counter_;
    surfaces_2d_.push_back({2, tag});
    return tag;
}

std::vector<std::pair<int,int>> StubGmshBackend::boolean_cut(
    const std::vector<std::pair<int,int>>& objects,
    const std::vector<std::pair<int,int>>&, bool)
{
    return objects; // stub: unchanged
}

std::vector<std::pair<int,int>> StubGmshBackend::boolean_fragment(
    const std::vector<std::pair<int,int>>& objects,
    const std::vector<std::pair<int,int>>& tools)
{
    auto r = objects;
    r.insert(r.end(), tools.begin(), tools.end());
    return r;
}

int StubGmshBackend::add_physical_group(int dim, const std::vector<int>& tags,
                                         const std::string& name, int tag) {
    int t = (tag >= 0) ? tag : ++phys_group_tag_counter_;
    phys_groups_.push_back({dim, tags, name, t});
    return t;
}

int StubGmshBackend::add_math_eval_field(const std::string&) {
    return ++field_counter_;
}

int StubGmshBackend::add_constant_field(double, const std::vector<int>&) {
    return ++field_counter_;
}

void StubGmshBackend::set_background_field(int field_tag) {
    background_field_ = field_tag;
}

void StubGmshBackend::generate_mesh(int) { mesh_generated_ = true; }

void StubGmshBackend::write_mesh(const std::string& filename) {
    last_write_path_ = filename;
}

std::vector<std::pair<int,int>> StubGmshBackend::get_entities_2d() {
    return surfaces_2d_;
}

// ─── Factory ──────────────────────────────────────────────────────────────────

std::unique_ptr<IGmshBackend> make_default_backend() {
#ifdef STATOR_WITH_GMSH
    extern std::unique_ptr<IGmshBackend> make_real_gmsh_backend();
    return make_real_gmsh_backend();
#else
    std::cerr << "[stator] Warning: STATOR_WITH_GMSH not defined; using StubGmshBackend.\n";
    return std::make_unique<StubGmshBackend>();
#endif
}

} // namespace stator

File: stator_pipeline/gmsh_backend.py
"""gmsh_backend.py — Abstract GMSH backend and in-memory stub."""
from __future__ import annotations

import math
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Optional

MAX_SLOTS = 256
MAX_PHYS_GROUPS = 512
MAX_TAGS_PER_GROUP = 256


class GmshBackend(ABC):
    """Abstract base for GMSH geometry/mesh backends."""

    @abstractmethod
    def initialize(self, model_name: str) -> None: ...
    @abstractmethod
    def synchronize(self) -> None: ...
    @abstractmethod
    def finalize(self) -> None: ...
    @abstractmethod
    def set_option(self, name: str, value: float) -> None: ...
    @abstractmethod
    def add_point(self, x: float, y: float, z: float, mesh_size: float) -> int: ...
    @abstractmethod
    def add_line(self, start: int, end: int) -> int: ...
    @abstractmethod
    def add_circle(self, cx: float, cy: float, cz: float, r: float) -> int: ...
    @abstractmethod
    def add_arc(self, start: int, centre: int, end: int) -> int: ...
    @abstractmethod
    def add_curve_loop(self, tags: list[int]) -> int: ...
    @abstractmethod
    def add_plane_surface(self, loop_tags: list[int]) -> int: ...
    @abstractmethod
    def boolean_cut(self, objects: list[tuple[int, int]], tools: list[tuple[int, int]], remove_tool: bool = False) -> list[tuple[int, int]]: ...
    @abstractmethod
    def boolean_fragment(self, objects: list[tuple[int, int]], tools: list[tuple[int, int]]) -> list[tuple[int, int]]: ...
    @abstractmethod
    def add_physical_group(self, dim: int, tags: list[int], name: str, tag: int = -1) -> int: ...
    @abstractmethod
    def add_math_eval_field(self, expr: str) -> int: ...
    @abstractmethod
    def add_constant_field(self, value: float, surfaces: list[int]) -> int: ...
    @abstractmethod
    def set_background_field(self, field_tag: int) -> None: ...
    @abstractmethod
    def generate_mesh(self, dim: int) -> None: ...
    @abstractmethod
    def write_mesh(self, filename: str) -> None: ...
    @abstractmethod
    def get_entities_2d(self) -> list[tuple[int, int]]: ...


@dataclass
class PhysGroupRecord:
    dim: int
    tags: list[int]
    name: str
    tag: int


class StubGmshBackend(GmshBackend):
    """In-memory stub that tracks all operations without calling GMSH."""

    def __init__(self) -> None:
        self._initialized = False
        self._finalized = False
        self._sync_count = 0
        self._point_counter = 0
        self._line_counter = 0
        self._curve_loop_counter = 0
        self._surface_counter = 0
        self._field_counter = 0
        self._phys_group_tag_counter = 1000
        self._surfaces_2d: list[tuple[int, int]] = []
        self._phys_groups: list[PhysGroupRecord] = []
        self._background_field: int = -1
        self._mesh_generated = False
        self._last_write_path = ""

    def initialize(self, model_name: str) -> None:
        self._initialized = True

    def synchronize(self) -> None:
        self._sync_count += 1

    def finalize(self) -> None:
        self._finalized = True

    def set_option(self, name: str, value: float) -> None:
        pass

    def add_point(self, x: float, y: float, z: float, mesh_size: float) -> int:
        self._point_counter += 1
        return self._point_counter

    def add_line(self, start: int, end: int) -> int:
        self._line_counter += 1
        return self._line_counter

    def add_circle(self, cx: float, cy: float, cz: float, r: float) -> int:
        self._line_counter += 1
        return self._line_counter

    def add_arc(self, start: int, centre: int, end: int) -> int:
        self._line_counter += 1
        return self._line_counter

    def add_curve_loop(self, tags: list[int]) -> int:
        self._curve_loop_counter += 1
        return self._curve_loop_counter

    def add_plane_surface(self, loop_tags: list[int]) -> int:
        self._surface_counter += 1
        tag = self._surface_counter
        self._surfaces_2d.append((2, tag))
        return tag

    def boolean_cut(self, objects: list[tuple[int, int]], tools: list[tuple[int, int]], remove_tool: bool = False) -> list[tuple[int, int]]:
        return list(objects)

    def boolean_fragment(self, objects: list[tuple[int, int]], tools: list[tuple[int, int]]) -> list[tuple[int, int]]:
        return list(objects) + list(tools)

    def add_physical_group(self, dim: int, tags: list[int], name: str, tag: int = -1) -> int:
        if tag < 0:
            self._phys_group_tag_counter += 1
            tag = self._phys_group_tag_counter
        self._phys_groups.append(PhysGroupRecord(dim=dim, tags=list(tags), name=name, tag=tag))
        return tag

    def add_math_eval_field(self, expr: str) -> int:
        self._field_counter += 1
        return self._field_counter

    def add_constant_field(self, value: float, surfaces: list[int]) -> int:
        self._field_counter += 1
        return self._field_counter

    def set_background_field(self, field_tag: int) -> None:
        self._background_field = field_tag

    def generate_mesh(self, dim: int) -> None:
        self._mesh_generated = True

    def write_mesh(self, filename: str) -> None:
        self._last_write_path = filename

    def get_entities_2d(self) -> list[tuple[int, int]]:
        return list(self._surfaces_2d)


def make_default_backend() -> GmshBackend:
    """Return the default (stub) backend. Replace with real GMSH when available."""
    return StubGmshBackend()
